bindings/python/iree/runtime/hal.cc

```diff
@@ -7,10 +7,194 @@
+// HalAllocator
+//------------------------------------------------------------------------------
+
+py::dict HalAllocator::QueryStatistics() {
+  py::dict items;
+  iree_hal_allocator_statistics_t stats;
+  iree_hal_allocator_query_statistics(raw_ptr(), &stats);
+#if IREE_STATISTICS_ENABLE
+  items["host_bytes_peak"] = stats.host_bytes_peak;
+  items["host_bytes_allocated"] = stats.host_bytes_allocated;
+  items["host_bytes_freed"] = stats.host_bytes_freed;
+  items["device_bytes_peak"] = stats.device_bytes_peak;
+  items["device_bytes_allocated"] = stats.device_bytes_allocated;
+  items["device_bytes_freed"] = stats.device_bytes_freed;
+#endif
+  return items;
+}
+
+py::str HalAllocator::FormattedStatistics() {
+  // Perform all allocating string manipulation without early exit.
+  iree_string_builder_t builder;
+  iree_string_builder_initialize(iree_allocator_system(), &builder);
+  iree_hal_allocator_statistics_t stats;
+  iree_hal_allocator_query_statistics(raw_ptr(), &stats);
+  auto status = iree_hal_allocator_statistics_format(&stats, &builder);
+  iree_string_view_t view = iree_string_builder_view(&builder);
+  py::str result = py::str(view.data, view.size);
+  iree_string_builder_deinitialize(&builder);
+
+  // Check/raise after all memory alloc/dealloc.
+  CheckApiStatus(status, "unable to format statistics");
+  return result;
+}
+
+py::object HalAllocator::AllocateBufferCopy(
+    int memory_type, int allowed_usage, py::object buffer,
+    std::optional<iree_hal_element_types_t> element_type) {
+  // Request a view of the buffer (use the raw python C API to avoid
+  // some allocation and copying at the pybind level).
+  Py_buffer py_view;
+  // Note that only C-Contiguous ND-arrays are presently supported, so
+  // only request that via PyBUF_ND. Long term, we should consult an
+  // "oracle" in the runtime to determine the precise required format
+  // and set flags accordingly (and fallback/copy on failure).
+  int flags = PyBUF_FORMAT | PyBUF_ND;
+
+  // Acquire the backing buffer and setup RAII release.
+  if (PyObject_GetBuffer(buffer.ptr(), &py_view, flags) != 0) {
+    // The GetBuffer call is required to set an appropriate error.
+    throw py::error_already_set();
+  }
+  PyBufferReleaser py_view_releaser(py_view);
+
+  iree_hal_buffer_t* hal_buffer;
+  CheckApiStatus(
+      iree_hal_allocator_allocate_buffer(
+          raw_ptr(), memory_type, allowed_usage, py_view.len,
+          iree_make_const_byte_span(py_view.buf, py_view.len), &hal_buffer),
+      "Failed to allocate device visible buffer");
+
+  if (!element_type) {
+    return py::cast(HalBuffer::StealFromRawPtr(hal_buffer),
+                    py::return_value_policy::move);
+  }
+
+  // Create the buffer_view. (note that numpy shape is ssize_t, so we need to
+  // copy).
+  iree_hal_encoding_type_t encoding_type =
+      IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR;
+  std::vector<int> dims(py_view.ndim);
```

Comment:
you'll need to use `iree_hal_dim_t` instead of int - intent is that it can be mapped to size_t or forced 32/64 so you'll probably always need this copy to adapt